# StrucVar_vcf2report
This repository hosts the scripts used in the Centre of Molecular Pathology to generate a filtered tab delimited structural variant (SV) report from VCF output generated by Manta and/or Pindel. The NGS data is generated through the Molecular Diagnostics Information Management System, a web application hosting an NGS pipeline for SNV, CNV and SV calling - developed by the bioinformatics team at the Centre of Molecular Pathology in Sutton, UK. 

## Usage & Installation
It does not require any installation other than cloning this repository. 

For run usage:

`python sv_vcf2report.py --help`

## Dependencies
The script is configured to handle VCF (currently VCF v4.1) output generated by Manta and/or Pindel.

*Software dependencies*

The main script takes the SV annotation ouput from AnnotSV (https://github.com/lgmgeo/AnnotSV) as one of its parameters while AnnotSV is run via TCL.

- `AnnotSV v2.0`
- `tcl v8.7`

---

*Python dependancies*
- `python 3.8.3`
- `pyvcf v0.6.8`
- `pysam v0.15.4`
- `numpy v1.18.1`

## Credits
Author: Sabri Jamal, Centre of Molecular Pathology

